### ISSUE TITLE
ROCM / HIP: Conv3d forward is very slow for some shapes.

### ISSUE THREAD

**Issue by IMbackK** [pytorch/pytorch#146998]
### 🐛 Describe the bug

With some conv3d shapes, such as those used by video models like hunyuan and moch1 are very slow on MIOpen backed platforms.

Below is a benchmark repducer:

```
import torch
import time

configs = [
	[128, 128, 3, 1],
	[256, 256, 3, 1],
	[512, 512, 3, 1],
	[128, 256, 1, 1],
	[512, 512, 3, (2, 2, 2)],
	[256, 256, 3, (2, 2, 2)],
	[128, 3, 3, 1]
]

inputs = [
	[1, 128, 67, 258, 258],
	[1, 256, 35, 130, 130],
	[1, 512, 35, 130, 130],
	[1, 128, 67, 258, 258],
	[1, 512, 35, 130, 130],
	[1, 256, 27, 258, 258],
	[1, 128, 67, 258, 258],
]


def conv3dbenchmark(configs: list[list[int]], inputs: list[list[int]], repeat: int, dtype: torch.dtype, device: torch.device):
	modules = list()
	assert len(inputs) == len(configs)

	for config in configs:
		modules.append(torch.nn.Conv3d(config[0], config[1], config[2], stride=config[3]).to(device, dtype))

	for i in range(len(modules)):
		x = torch.randn(inputs[i]).to(device, dtype)
		print(f"Running Conv3d config: {configs[i]} input: {inputs[i]} type: {dtype}")
		start = time.perf_counter()
		for n in range(repeat):
			modules[i].forward(x)
		torch.cuda.synchronize(device)
		print(f"Time {(time.perf_counter() - start) / repeat} seconds\n")


if __name__ == "__main__":
	device = torch.device(0)

	conv3dbenchmark(configs, inputs, 5, torch.bfloat16, device)
	conv3dbenchmark(configs, inputs, 5, torch.float16, device)
```

The benchmark was always run twice to allow miopen to cache its solutions, however using MIOPEN_FIND_MODE=2 provides equivalent performance without the need to cache any solutions.

[mi100.txt](https://github.com/user-attachments/files/18771827/mi100.txt)

[3090.txt](https://github.com/user-attachments/files/18771828/3090.txt)

[rx6800xt.txt](https://github.com/user-attachments/files/18771829/rx6800xt.txt)

[cpu.txt](https://github.com/user-attachments/files/18771857/cpu.txt)

Comparison was done to the 3090 which was choosen as a device with roughly similar raw compute and memory bandwidth as mi100.
Additionally an epic 7552 was used as another point of comparison.

It can be seen that when comparing the mi100, on most configurations, such as `config: [512, 512, 3, 1] input: [1, 512, 35, 130, 130] type: torch.float16` the cuDNN device holds a 10x advantage.
The performance on RX 6800XT can only be described as broken, with it often performing much worse than the cpu. 

In addition to the above micro benchmark a benchmark executing a VAE decode is provided here: https://uvos.xyz/git/uvos/HyDecodeRepo
On 3090 this takes about 25 Seconds, while on MI100 120 seconds are required.
Pytorch Profiler Traces showing the slected kernels are given below:

[hunyuan_vae_decode_rtx_3090.zip](https://github.com/user-attachments/files/18772005/hunyuan_vae_decode_rtx_3090.zip)

[hunyuan_vae_decode_rtx_MI100.zip](https://github.com/user-attachments/files/18772027/hunyuan_vae_decode_rtx_MI100.zip)

### Versions

MIOpen 6.2.4
Pytorch 2.6.0

cc @msaroufim @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd

**Comment by naromero77amd** [comment]
Hi @IMbackK, Thanks for putting together that benchmark reproducer.

There were some recent changes in PyTorch upstream:
https://github.com/pytorch/pytorch/pull/145294

Could you please use the nightly PyTorch wheel and re-run the benchmark with:
```
torch.backends.cudnn.benchmark=true
```

This should do an exhaustive search and update any DBs.

**Comment by IMbackK** [comment]
Hi @naromero77amd, 

I just tried 2.7.0.dev20250213+rocm6.3 with a pass with torch.backends.cudnn.benchmark=true, unfortunately this dose not change the results any: 

MI100:
```
Running Conv3d config: [128, 128, 3, 1] input: [1, 128, 67, 258, 258] type: torch.bfloat16
Time 0.44905592760187574 seconds

Running Conv3d config: [256, 256, 3, 1] input: [1, 256, 35, 130, 130] type: torch.bfloat16
Time 0.20562958880327642 seconds

Running Conv3d config: [512, 512, 3, 1] input: [1, 512, 35, 130, 130] type: torch.bfloat16
Time 0.827823149797041 seconds

Running Conv3d config: [128, 256, 1, 1] input: [1, 128, 67, 258, 258] type: torch.bfloat16
Time 0.044789208797737956 seconds

Running Conv3d config: [512, 512, 3, (2, 2, 2)] input: [1, 512, 35, 130, 130] type: torch.bfloat16
Time 0.033959378401050344 seconds

Running Conv3d config: [256, 256, 3, (2, 2, 2)] input: [1, 256, 27, 258, 258] type: torch.bfloat16
Time 0.03877393120201304 seconds

Running Conv3d config: [128, 3, 3, 1] input: [1, 128, 67, 258, 258] type: torch.bfloat16
Time 0.16032897619879805 seconds

Running Conv3d config: [128, 128, 3, 1] input: [1, 128, 67, 258, 258] type: torch.float16
Time 0.3853755710006226 seconds

Running Conv3d config: [256, 256, 3, 1] input: [1, 256, 35, 130, 130] type: torch.float16
Time 0.19432329660048708 seconds

Running Conv3d config: [512, 512, 3, 1] input: [1, 512, 35, 130, 130] type: torch.float16
Time 0.7843209019978531 seconds

Running Conv3d config: [128, 256, 1, 1] input: [1, 128, 67, 258, 258] type: torch.float16
Time 0.052972708002198485 seconds

Running Conv3d config: [512, 512, 3, (2, 2, 2)] input: [1, 512, 35, 130, 130] type: torch.float16
Time 0.032184987200889734 seconds

Running Conv3d config: [256, 256, 3, (2, 2, 2)] input: [1, 256, 27, 258, 258] type: torch.float16
Time 0.03536759900161997 seconds

Running Conv3d config: [128, 3, 3, 1] input: [1, 128, 67, 258, 258] type: torch.float16
Time 0.13743250500410795 seconds
```

It seams the viable miopen kernels are just not very efficient for these cases.

**Comment by JoshuaBowerman** [comment]
I'm also seeing similar(even worse) results with a 7900XTX. 
```Running Conv3d config: [128, 128, 3, 1] input: [1, 128, 67, 258, 258] type: torch.bfloat16
Time 44.5200458928 seconds

Running Conv3d config: [256, 256, 3, 1] input: [1, 256, 35, 130, 130] type: torch.bfloat16
Time 19.448282552399995 seconds

Running Conv3d config: [512, 512, 3, 1] input: [1, 512, 35, 130, 130] type: torch.bfloat16
Time 70.03178845359999 seconds

Running Conv3d config: [128, 256, 1, 1] input: [1, 128, 67, 258, 258] type: torch.bfloat16
Time 2.406099089999998 seconds

Running Conv3d config: [512, 512, 3, (2, 2, 2)] input: [1, 512, 35, 130, 130] type: torch.bfloat16
Time 4.970142854599999 seconds

Running Conv3d config: [256, 256, 3, (2, 2, 2)] input: [1, 256, 27, 258, 258] type: torch.bfloat16
Time 4.452221212600034 seconds

Running Conv3d config: [128, 3, 3, 1] input: [1, 128, 67, 258, 258] type: torch.bfloat16
Time 19.986986238999997 seconds

Running Conv3d config: [128, 128, 3, 1] input: [1, 128, 67, 258, 258] type: torch.float16
Time 44.54232813780004 seconds

Running Conv3d config: [256, 256, 3, 1] input: [1, 256, 35, 130, 130] type: torch.float16
Time 19.74684859680001 seconds

Running Conv3d config: [512, 512, 3, 1] input: [1, 512, 35, 130, 130] type: torch.float16
Time 71.12273250160001 seconds

Running Conv3d config: [128, 256, 1, 1] input: [1, 128, 67, 258, 258] type: torch.float16
Time 2.417141667199985 seconds

Running Conv3d config: [512, 512, 3, (2, 2, 2)] input: [1, 512, 35, 130, 130] type: torch.float16
Time 5.025415738000038 seconds

Running Conv3d config: [256, 256, 3, (2, 2, 2)] input: [1, 256, 27, 258, 258] type: torch.float16
Time 4.51895615479998 seconds

Running Conv3d config: [128, 3, 3, 1] input: [1, 128, 67, 258, 258] type: torch.float16
Time 20.839066082800038 seconds```

**Comment by IMbackK** [comment]
Outch, Is that the same when run again? that looks so bad that i hope its just miopen caching its selections.
Altho my RDNA2 results are similar, even when miopen has cached its selections.

**Comment by JoshuaBowerman** [comment]
Unfortunately that's run 4 or 5. I'm also seeing similar issues with some Conv1d inputs. Though I don't have a Nvidia GPU to compare my results with.

**Comment by IMbackK** [comment]
Thats completely terrible, your better off using the cpu by like 30x.

**Comment by naromero77amd** [comment]
Thank you for reporting these performance problems.

The AMD MIOpen team is aware of these issues.

We don't have a timeline for improvements to these convolution uses on MI100 and Navi.

**Comment by IMbackK** [comment]
Could we have an issue opened against miopen? I think this is necessary for visibility.
I think this performance deficit, even vs cpu execution should be at least documented.
The users of MIOpen can get order of magnitude improvements by simply transferring operations to the cpu, they should be made aware of this.

**Comment by naromero77amd** [comment]
@IMbackK I think it's fine if you want to open an issue here: https://github.com/ROCm/MIOpen/issues.

It would raise awareness of the problem. You can include a link to this PyTorch issue as well.
